R/method-differentialAbundance.R

```diff
@@ -76,7 +76,36 @@ cleanComparatorVariable <- function(data, comparator, verbose = c(TRUE, FALSE))
     return(data)
 }
 
-# TODO update plugin
+DifferentialAbundanceResult <- setClass("DifferentialAbundanceResult", representation(
+    effectSizeLabel = 'character',
+    statistics = 'data.frame'
+), prototype = prototype(
+    effectSizeLabel = 'log2(Fold Change)',
+    statistics = data.frame(effectSize = numeric(0),
+                            pValue = numeric(0),
+                            adjustedPValue = numeric(0),
+                            pointID = character(0))
+))
+
+setGeneric("toJSON",
+  function(object, ...) standardGeneric("toJSON"),
+  signature = "object"
+)
+
+setMethod("toJSON", signature("DifferentialAbundanceResult"), function(object, ...) {
+  tmp <- character()
+
+  tmp <- paste0('"effectSizeLabel": ', jsonlite::toJSON(jsonlite::unbox(object@effectSizeLabel)), ',')
+  tmp <- paste0(tmp, paste0('"statistics": ', jsonlite::toJSON(object@statistics)))
+
+  tmp <- paste0("{", tmp, "}")
+  return(tmp)
+})
+
+# these let jsonlite::toJSON work by using the custom toJSON method for our custom result class
+asJSONGeneric <- getGeneric("asJSON", package = "jsonlite")
+setMethod(asJSONGeneric, "DifferentialAbundanceResult", function(x, ...) toJSON(x))
```

Comment:
interesting. So in other words if i call `asJSON` on a diffabund result, what will actually happen is that our special `toJSON` method will get called? Do i understand that right?

Reply:
But later we call `toJSON`, which uses regular dispatch since above we've defined it for the diff abund result signature. Where do we use `asJSON`?

Reply:
AsJson is internal to jsonlite::toJSON, which we're using in writeStats and I didn't want to change that.